### PR TITLE
Base data::Dataset class

### PR DESCRIPTION
**Pull request by yxlao**
- Base `data::Dataset` class. With mechanism to set data root or 
- `utility::filesystem::GetHomeDirectory()`
- `utility::StringStartsWith()`
- `utility::StringEndsWith()`

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4258)

### CODE REVIEW

**Review by reyanshsolis — APPROVED** [review]
LGTM

**Review by ssheorey — COMMENTED** [review]
Could you add a simple dataset class to give a sense of how this will be used? We can start with the any data in the `examples/test_data` folder (e.g. fountain_small or RGBD) so that download / extract logic is not needed.

File: cpp/open3d/data/Dataset.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+/// dataset classes. The downloaded data will be stored in the Open3D's data
+/// root directory.
+///
+/// - A dataset class locates the data root directory in the following order:
+///   (a) User-specified by `data_root` when instantiating a dataset object.
+///   (b) OPEN3D_DATA_ROOT environment variable.
+///   (c) $HOME/open3d_data.
+///   By default, (c) will be used, and it is also the recommended way.
+/// - When a dataset object is instantiated, the corresponding data will be
+///   downloaed and extracted. If the data already exists and the checksum
```
[the remark this answers is not in the document]
Comment (yxlao) [reply cpp/open3d/data/Dataset.h:48]: Done

**Comment by yxlao** [comment]
> Could you add a simple dataset class to give a sense of how this will be used? We can start with the any data in the `examples/test_data` folder (e.g. fountain_small or RGBD) so that download / extract logic is not needed.

That's a good idea. I will do it in the next PR since I will expect the code size to be as big as this PR.

**Pull request merged**